File: bot/domain/repositories/playlist_repository.py
from __future__ import annotations
import json
import os
import shutil
from pathlib import Path
from typing import List, Optional, Dict, Any

from ..entities.playlist import Playlist
from ...pkg.logger import logger


class PlaylistRepository:
    """Repository for persisting playlists to JSON files"""

    def __init__(self, base_path: str = "playlist"):
        self.base_path = Path(base_path)
        self.base_path.mkdir(exist_ok=True)

    def _get_file_path(self, playlist_name: str) -> Path:
        """Get file path for playlist"""
        safe_name = "".join(
            c if c.isalnum() or c in ("-", "_") else "_" for c in playlist_name
        )
        return self.base_path / f"{safe_name}.json"

    def save(self, playlist: Playlist) -> bool:
        """Save playlist to file with atomic write"""
        try:
            file_path = self._get_file_path(playlist.name)
            
            # Create backup if file exists
            if file_path.exists():
                backup_path = file_path.with_suffix('.json.backup')
                try:
                    shutil.copy2(file_path, backup_path)
                except Exception as backup_error:
                    logger.warning(f"Could not create backup: {backup_error}")
            
            # Atomic write using temp file
            temp_path = file_path.with_suffix('.json.tmp')
            with open(temp_path, "w", encoding="utf-8") as f:
                json.dump(playlist.to_dict(), f, indent=2, ensure_ascii=False)
            
            # Rename for atomicity
            temp_path.replace(file_path)
            
            logger.info(f"✅ Saved playlist: {playlist.name}")
            return True
        except PermissionError as e:
            logger.error(f"❌ Permission denied saving playlist {playlist.name}: {e}")
            return False
        except OSError as e:
            logger.error(f"❌ Disk error saving playlist {playlist.name}: {e}")
            return False
        except Exception as e:
            logger.exception(f"❌ Unexpected error saving playlist {playlist.name}")
            return False

    def load(self, playlist_name: str) -> Optional[Playlist]:
        """Load playlist from file"""
        try:
            file_path = self._get_file_path(playlist_name)
            if not file_path.exists():
                logger.debug(f"Playlist file not found: {playlist_name}")
                return None

            with open(file_path, "r", encoding="utf-8") as f:
                data = json.load(f)
            
            logger.info(f"✅ Loaded playlist: {playlist_name}")
            return Playlist.from_dict(data)
        except json.JSONDecodeError as e:
            logger.error(f"❌ Invalid JSON in playlist {playlist_name}: {e}")
            return None
        except OSError as e:
            logger.error(f"❌ Error reading playlist {playlist_name}: {e}")
            return None
        except Exception as e:
            logger.exception(f"❌ Unexpected error loading playlist {playlist_name}")
            return None

    def delete(self, playlist_name: str) -> bool:
        """Delete playlist file"""
        try:
            file_path = self._get_file_path(playlist_name)
            if not file_path.exists():
                logger.warning(f"Playlist file not found: {playlist_name}")
                return False
            
            # Move to trash instead of permanent delete
            trash_path = file_path.with_suffix('.json.deleted')
            file_path.rename(trash_path)
            
            logger.info(f"🗑️ Deleted playlist: {playlist_name}")
            return True
        except OSError as e:
            logger.error(f"❌ Error deleting playlist {playlist_name}: {e}")
            return False
        except Exception as e:
            logger.exception(f"❌ Unexpected error deleting playlist {playlist_name}")
            return False

    def list_all(self) -> List[str]:
        """List all available playlists"""
        try:
            playlist_names = []
            for file_path in self.base_path.glob("*.json"):
                try:
                    with open(file_path, "r", encoding="utf-8") as f:
                        data = json.load(f)
                    playlist_names.append(data["name"])
                except Exception as e:
                    logger.warning(f"Could not read playlist {file_path}: {e}")
                    continue
            return sorted(playlist_names)
        except Exception as e:
            logger.error(f"❌ Error listing playlists: {e}")
            return []

    def exists(self, playlist_name: str) -> bool:
        """Check if playlist exists"""
        return self._get_file_path(playlist_name).exists()

File: internal/infrastructure/persistence/playlist_repository.go
package persistence

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"
	"sync"

	"github.com/vuongmanhnghia/discord-music-bot/internal/domain/entities"
)

// PlaylistRepository handles persistence of playlists to JSON files
type PlaylistRepository struct {
	basePath string
	mu       sync.RWMutex
}

// NewPlaylistRepository creates a new playlist repository
func NewPlaylistRepository(basePath string) (*PlaylistRepository, error) {
	// Ensure base path exists
	if err := os.MkdirAll(basePath, 0755); err != nil {
		return nil, fmt.Errorf("failed to create playlist directory: %w", err)
	}

	return &PlaylistRepository{
		basePath: basePath,
	}, nil
}

// Save saves a playlist to a JSON file with atomic write
func (r *PlaylistRepository) Save(playlist *entities.Playlist) error {
	r.mu.Lock()
	defer r.mu.Unlock()

	filePath := r.getFilePath(playlist.Name)

	// Create backup if file exists
	if _, err := os.Stat(filePath); err == nil {
		backupPath := filePath + ".backup"
		if err := r.copyFile(filePath, backupPath); err != nil {
			// Log warning but continue
			fmt.Printf("Warning: could not create backup: %v\n", err)
		}
	}

	// Atomic write using temp file
	tempPath := filePath + ".tmp"
	file, err := os.Create(tempPath)
	if err != nil {
		return fmt.Errorf("failed to create temp file: %w", err)
	}

	encoder := json.NewEncoder(file)
	encoder.SetIndent("", "  ")
	if err := encoder.Encode(playlist); err != nil {
		file.Close()
		os.Remove(tempPath)
		return fmt.Errorf("failed to encode playlist: %w", err)
	}

	if err := file.Close(); err != nil {
		os.Remove(tempPath)
		return fmt.Errorf("failed to close temp file: %w", err)
	}

	// Rename for atomicity
	if err := os.Rename(tempPath, filePath); err != nil {
		return fmt.Errorf("failed to rename temp file: %w", err)
	}

	return nil
}

// Load loads a playlist from a JSON file
func (r *PlaylistRepository) Load(playlistName string) (*entities.Playlist, error) {
	r.mu.RLock()
	defer r.mu.RUnlock()

	filePath := r.getFilePath(playlistName)

	file, err := os.Open(filePath)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil // Playlist not found
		}
		return nil, fmt.Errorf("failed to open playlist file: %w", err)
	}
	defer file.Close()

	var playlist entities.Playlist
	decoder := json.NewDecoder(file)
	if err := decoder.Decode(&playlist); err != nil {
		return nil, fmt.Errorf("failed to decode playlist: %w", err)
	}

	return &playlist, nil
}

// Delete deletes a playlist file (soft delete by renaming)
func (r *PlaylistRepository) Delete(playlistName string) error {
	r.mu.Lock()
	defer r.mu.Unlock()

	filePath := r.getFilePath(playlistName)

	if _, err := os.Stat(filePath); os.IsNotExist(err) {
		return fmt.Errorf("playlist not found: %s", playlistName)
	}

	// Soft delete by renaming
	deletedPath := filePath + ".deleted"
	if err := os.Rename(filePath, deletedPath); err != nil {
		return fmt.Errorf("failed to delete playlist: %w", err)
	}

	return nil
}

// Exists checks if a playlist exists
func (r *PlaylistRepository) Exists(playlistName string) bool {
	r.mu.RLock()
	defer r.mu.RUnlock()

	filePath := r.getFilePath(playlistName)
	_, err := os.Stat(filePath)
	return err == nil
}

// ListAll returns all playlist names
func (r *PlaylistRepository) ListAll() ([]string, error) {
	r.mu.RLock()
	defer r.mu.RUnlock()

	files, err := os.ReadDir(r.basePath)
	if err != nil {
		return nil, fmt.Errorf("failed to read playlist directory: %w", err)
	}

	playlists := make([]string, 0)
	for _, file := range files {
		if file.IsDir() {
			continue
		}

		name := file.Name()
		if strings.HasSuffix(name, ".json") &&
			!strings.HasSuffix(name, ".backup") &&
			!strings.HasSuffix(name, ".tmp") &&
			!strings.HasSuffix(name, ".deleted") {
			// Remove .json extension
			playlistName := strings.TrimSuffix(name, ".json")
			playlists = append(playlists, playlistName)
		}
	}

	return playlists, nil
}

// getFilePath returns the full file path for a playlist
func (r *PlaylistRepository) getFilePath(playlistName string) string {
	// Sanitize filename
	safeName := r.sanitizeFilename(playlistName)
	return filepath.Join(r.basePath, safeName+".json")
}

// sanitizeFilename removes unsafe characters from filename
func (r *PlaylistRepository) sanitizeFilename(name string) string {
	// Replace unsafe characters with underscore
	safeName := ""
	for _, char := range name {
		if (char >= 'a' && char <= 'z') ||
			(char >= 'A' && char <= 'Z') ||
			(char >= '0' && char <= '9') ||
			char == '-' || char == '_' {
			safeName += string(char)
		} else {
			safeName += "_"
		}
	}
	return safeName
}

// copyFile copies a file from src to dst
func (r *PlaylistRepository) copyFile(src, dst string) error {
	sourceFile, err := os.Open(src)
	if err != nil {
		return err
	}
	defer sourceFile.Close()

	destFile, err := os.Create(dst)
	if err != nil {
		return err
	}
	defer destFile.Close()

	_, err = io.Copy(destFile, sourceFile)
	return err
}
